Task1

Maxoriy opened the pull request.

Maxoriy committed:
switch to maven

Maxoriy committed:
Fixed the resource leak and limited the number of messages in the log to m_iQueueLength

Maxoriy committed:
added the localization

Maxoriy committed:
Divided the creation method into simple methods.
Added a menu item that allows you to close the application.
Added localization to the methods.

alklepin commented on src/main/java/gui/GameWindow.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.ResourceBundle;
+
+public class GameWindow extends JInternalFrame {
+    public GameWindow(ResourceBundle bundle, int width, int height) {
```
Вот тут тоже обозначу проблему. Для локализации теперь приходится протаскивать ResourceBundle как отдельный параметр, что ведет к замусориванию интерфейса класса (ну ведь действительно, ресурсы как бы не являются параметром этого класса). Поэтому, по идее, при работе с ресурсами связь с ними делают или через статическое связывание (синглтон с ресурсами, к которому напрямую через статический метод обращается использующий его класс), или через более продвинутые шаблоны Service Locator или Dependency Injection (эти конструкции будем еще обсуждать в нашем курсе). То есть вот тут увидели, чем эти связывающие шаблоны помогают с точки зрения улучшения кода.
В принципе, можно было бы попробовать и здесь что-то подобное применить, но, думаю, это следовало бы рассматривать уже как отдельную задачу с отдельными баллами.

alklepin commented on src/main/java/gui/GameWindow.java:
```diff
@@ -0,0 +1,19 @@
+        panel.add(m_visualizer, BorderLayout.CENTER);
+        getContentPane().add(panel);
+        pack();
+        super.setSize(width, height);
```
Зачем здесь super? Обычно такая конструкция используется внутри переопределяемого метода, чтобы явно вызвать унаследованную реализацию... Здесь же точно не такой случай.

alklepin commented on src/main/java/gui/MainApplicationFrame.java:
```diff
@@ -0,0 +1,130 @@
+        LogWindow logWindow = createLogWindow();
+        addWindow(logWindow);
+
+        addWindow(new GameWindow(bundle, 400, 400));
```
А зачем установку размера переместили внутрь конструктора? Это не ошибка, но интересно услышать аргументацию...

Maxoriy replied:
Код становится более читаемым и понятным. Три строчки превратились в одну :) (Только из-за этого)
Также в новом изменении, я сразу написал addWindow(createLogWindow())

alklepin replied:
Тогда скажу так. Насчет читаемости - крайне спорный аргумент. Исходная версия читалась лучше: создали окно - установили размер. А теперь надо разбираться, что там в конструкторе происходит. К тому же, а если еще какие-то параметры захотим настраивать? Тоже в конструктор потащим их?
Так что на будущее посоветую не бороться за уменьшение числа строк. Не надо писать лишнего. А если написано по сути, то пытаться формально все в одну строку уместить - не лучший вариант.

alklepin commented on src/main/java/gui/MainApplicationFrame.java:
```diff
@@ -0,0 +1,130 @@
+                    JOptionPane.YES_NO_OPTION,
+                    JOptionPane.QUESTION_MESSAGE,
+                    null,
+                    new String[]{bundle.getString("yes"), bundle.getString("no")},
```
Для JOptionPane работает та же механика, что и для Swing в целом. Так что стоило не здесь локально выполнять локализацию, а сделать это глобально (собственно, в этом же и состояла задача?)

alklepin commented on src/main/java/gui/RobotsProgram.java:
```diff
@@ -0,0 +1,28 @@
+
+public class RobotsProgram {
+    public static void main(String[] args) {
+        ResourceBundle bundle = ResourceBundle.getBundle("ru");
```
С бандлами работают немного не так. Во-первых, у бандла есть имя. И именно его указывают при загрузке. Например:
```
ResourceBundle bundle = ResourceBundle.getBundle("my_resources");
```
И затем загружается бандл с учетом локализации. Для этого берется текущая локаль (допустим, русская, то есть ru_RU) и выполняется поиск файла с данными бандла в следующем порядке:
my_resources_ru_RU.properties
my_resources_ru.properties
my_resources.properties
И загружается первый найденный файл (точнее, будут загружены все найденные по такой схеме, и при поиске нужного ресурса они будут просматриваться в указанном порядке, пока ресурс не будет найден)
Надо поправить код, чтобы он соответствовал этой схеме использования (в частности, набор ресурсов выбираем не за счет указания имени бандла, а за счет установки локали).

alklepin commented:
На данный момент остаются две важные проблемы:
1. не решена задача, что окошко подтверждения выхода из приложения должно появляться на любых сценариях закрытия приложения (не только через меню)
2. не совсем правильно реализована работа с ресурсами (см. мой комментарий про правильное использование бандлов).
Для того, чтобы считать задачу решенной, надо все же эти две проблемы решить.

Maxoriy committed:
Fixed that the exit confirmation window did not appear in all closing scenarios.
Implemented the correct work with resources.
Removed the super in the GameWindow class.
Removed localization for the log.

Maxoriy committed:
Fixed method append in class LogWindowSource

alklepin commented on src/main/java/gui/MainApplicationFrame.java:
```diff
@@ -0,0 +1,142 @@
+                null);
+        if (confirm == JOptionPane.YES_OPTION) {
+            dispose();
+            System.exit(0);
```
Нет, так завершать приложение нельзя. Эта операция полностью останавливает виртуальную машину, что делает невозможным корректное завершение приложения. Надо сделать так, чтобы программа завершалась без вызова System.exit.

alklepin commented on src/main/java/gui/MainApplicationFrame.java:
```diff
@@ -0,0 +1,142 @@
+                JOptionPane.YES_NO_OPTION,
+                JOptionPane.QUESTION_MESSAGE,
+                null,
+                new String[]{"Да", "Нет"},
```
Так а почему этот код остался в таком виде, если уже сделана более правильная версия кода для локализации?

alklepin commented:
Для получения полного балла за эту задачу осталось только избавиться от System.exit.

Maxoriy committed:
returned the localization

Maxoriy committed:
Bye system.exit

Maxoriy commented:
На всякий случай ещё раз здесь напишу. От System.exit избавился в 3-й задаче

alklepin commented:
Можно ветку влить в мастер. :)
